anesthetic/utils.py

```diff
@@ -156,3 +157,9 @@ def unique(a):
         if x not in b:
             b.append(x)
     return b
+
+
+def kwargs_in_func(kwargs_dict, func):
+    """Find all kwargs that will be used by func."""
+    aspec = inspect.getargspec(func)
+    return {key: kwargs_dict[key] for key in kwargs_dict if key in aspec.args}
```

Comment:
This could be a really nice solution to the problem, however I'm concerned that matplotlib is not capable of inspecting it's keyword arguments like this (indeed general python code gives no such guarantees). For example, if you tried using this on one of the functions in anesthetic, it wouldn't be able to extract the kwargs that it are expected due to the 'pop' implementation. Does this code actually work for your plotting examples?

I think the only way to address this problem in the current paradigm is to have two additional kwargs to ``MCMCSamples.plot_2d``: ``1d_kwargs``, ``2d_kwargs``. These should be dictionaries, and indicate the arguments that should be passed *only* to the 1d or 2d plotting functions. The remaining *args and **kwargs should get passed to both (e.g. a color-based kwarg).

Reply:
I tried this with `plot`, `hist`, and `contour` and it seems to work. At least it catches all `args` of those functions. I did not make an exhaustive test of all possible `kwargs` though.

I don't think `1d_kwargs` and `2d_kwargs` would be solution we want, because `plot_1d` and `hist_1d` already have different `kwargs`.

Reply:
> I did not make an exhaustive test of all possible kwargs though.

The issue is that there are a lot of possible keyword arguments to the plotting functions, and an exhaustive test is not really feasible now, or for future changes to matplotlib/additions to anesthetic

> I don't think 1d_kwargs and 2d_kwargs would be the solution we want, because plot_1d and hist_1d already have different kwargs.

I don't think that would matter in practice. The kind of code I was imagining would be:
```python

1d_kwargs_temp = kwargs.pop('1d_kwargs',{})
1d_kwargs = kwargs.copy()
1d_kwargs.update(1d_kwargs_temp)
```
and then passing ``1d_kwargs`` to the relevant 1d function in place of ``kwargs``, independent of whether it's a ``hist`` or ``kde``

Reply:
(the above code would occur in ``MCMCSamples.plot_2d``, whilst ``MCMCSamples.plot_1d`` would not need such preprocessing)

Reply:
With regands the discussion above on how to implement ``types``, plot_2d should take 'upper_kwargs', 'lower_kwargs' and 'diagonal_kwargs' to pass on to the respective ``plot`` routine

Reply:
created `upper_kwargs`, `lower_kwargs` and `diagonal_kwargs` for `MCMCsamples.plot_2d`